File: test/integration/AliasableOnlyProbe.py
#===----------------------------------------------------------------------===#
#
# Part of the OpenCyphal project, under the MIT licence
# SPDX-License-Identifier: MIT
#
#===----------------------------------------------------------------------===#
#
# The C probe's reading, through the Python static-method accessors.
#
#===----------------------------------------------------------------------===#

import importlib
import struct
import sys


def main() -> int:
    package_root, package = sys.argv[1], sys.argv[2]
    sys.path.insert(0, package_root)
    pose = importlib.import_module(f"{package}.fixtures_aliasable.vendor.pose_1_0").Pose
    vec3 = importlib.import_module(f"{package}.fixtures_aliasable.vendor.vec3_1_0").Vec3

    buffer = memoryview(bytearray(struct.pack("<6f", 1.5, -2.25, 3.0, 4.0, 5.5, 6.75)))
    orientation = pose.get_orientation(buffer)
    y = vec3.get_y(orientation)
    set_result = vec3.set_z(buffer[:12], 9.5)
    z = vec3.get_z(buffer)
    short_read = vec3.get_z(buffer[:4])
    x = vec3.get_x(pose.get_position(buffer))

    ok = len(orientation) == 12 and y == 5.5 and set_result == 0 and z == 9.5 and short_read == 0 and x == 1.5
    print(
        f"aliasable-only Python: {'ok' if ok else 'FAILED'} (orientation {len(orientation)} bytes, y {y}, "
        f"set {set_result}, z {z}, short {short_read}, x {x})"
    )
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: test/integration/AliasableOnlyProbe.cpp
//===----------------------------------------------------------------------===//
//
// Part of the OpenCyphal project, under the MIT licence
// SPDX-License-Identifier: MIT
//
//===----------------------------------------------------------------------===//
//
// The C probe's reading, through the C++ static-member accessors, for both profiles.
//
//===----------------------------------------------------------------------===//

#include "fixtures_aliasable/vendor/Pose_1_0.hpp"
#include "fixtures_aliasable/vendor/Vec3_1_0.hpp"

#include <cstdio>
#include <cstring>
#include <span>

int main()
{
    const float  f[6] = {1.5f, -2.25f, 3.0f, 4.0f, 5.5f, 6.75f};
    std::uint8_t buffer[24];
    std::memcpy(buffer, f, sizeof buffer);

    using fixtures_aliasable::vendor::Pose;
    using fixtures_aliasable::vendor::Vec3;

    const std::span<const std::uint8_t> orientation = Pose::get_orientation(buffer);
    const float                         y           = Vec3::get_y(orientation);
    const std::int8_t                   set_result  = Vec3::set_z(std::span<std::uint8_t>(buffer, 12), 9.5f);
    const float                         z           = Vec3::get_z(std::span<const std::uint8_t>(buffer, 12));
    const float                         short_read  = Vec3::get_z(std::span<const std::uint8_t>(buffer, 4));
    const float                         x           = Vec3::get_x(Pose::get_position(buffer));

    const bool ok =
        orientation.size() == 12 && y == 5.5f && set_result == 0 && z == 9.5f && short_read == 0.0f && x == 1.5f;
    std::printf("aliasable-only C++: %s (orientation %zu bytes, y %g, set %d, z %g, short %g, x %g)\n",
                ok ? "ok" : "FAILED",
                orientation.size(),
                static_cast<double>(y),
                static_cast<int>(set_result),
                static_cast<double>(z),
                static_cast<double>(short_read),
                static_cast<double>(x));
    return ok ? 0 : 1;
}
